src/python/pants/backend/python/util_rules/pex.py

```diff
@@ -324,44 +368,55 @@ async def build_pex(
     if request.main is not None:
         argv.extend(request.main.iter_pex_args())
 
+    # TODO(John Sirois): Right now any request requirements will shadow corresponding pex path
+    #  requirements, which could lead to problems. Support shading python binaries.
+    #  See: https://github.com/pantsbuild/pants/issues/9206
+    if request.pex_path:
+        argv.extend(["--pex-path", ":".join(pex.name for pex in request.pex_path)])
+
     source_dir_name = "source_files"
     argv.append(f"--sources-directory={source_dir_name}")
+    sources_digest_as_subdir = await Get(
+        Digest, AddPrefix(request.sources or EMPTY_DIGEST, source_dir_name)
+    )
 
-    argv.extend(request.requirements)
+    additional_inputs_digest = request.additional_inputs or EMPTY_DIGEST
+    repository_pex_digest = (
+        request.repository_pex.digest if request.repository_pex else EMPTY_DIGEST
+    )
 
     constraint_file_digest = EMPTY_DIGEST
-    if request.apply_requirement_constraints and python_setup.requirement_constraints is not None:
+    if (
+        not request.is_lockfile and request.apply_requirement_constraints
```

Comment:
Should/does the constructor validate that only one of these is true at once? Alternatively, this seems like another case where a `union` more accurately represents which cases can be true simultaneously: `resolve_file: Lockfile | ConstraintsFile | None`.

Reply:
`apply_requirement_constraints` defaults to `True`. To avoid having to update lots of call sites, I kept that default for now. It felt redundant to set both `apply_requirement_constraints=False` and `is_lockfile=True`.

A union could make sense, although it's frustrating that Python doesn't have ADTs and we would need to define two new types. I'd like to avoid defining new types until we have more experience with the rest of the project.